background-script.js

```diff
@@ -11,62 +11,119 @@
+    }
+    return undefined;
+}
+async function getAllHeader(arr) {
+    if (Array.isArray(arr) && arr.length > 0) {
+        return arr;
+    }
+    return undefined;
+}
+function parseDisplayName(addr) {
+    let rv = parse5322.parseOneAddress(addr);
+    return {
+        name: rv.name,
+        email: rv.address,
+    }
+}
+// Find first text/plain body.
+function getBody(parts) {
+    // First check all parts in this level.
+    for (let part of parts) {
+        if (part.body && part.contentType == "text/plain") {
+            return part.body;
+        }
+    }
+    // Now check all subparts.
+    for (let part of parts) {
+        if (part.parts) {
+            let body = getBody(part.parts);
+            if (body) {
+                return body
+            }
+        }
     }
     return null;
 }
+// Compatibility layer for former experiment.
+async function getMsgData(messageId) {
+    let full = await browser.messages.getFull(messageId);
+    let date = await getFirstHeader(full.headers["date"]).then(
+        value => value ? new Date(value) : value
+    );
+    let from = await getAllHeader(full.headers["from"]).then(
+        value => value ? value.map(addr => parseDisplayName(addr)) : value
+    );
+    let replyTo = await getAllHeader(full.headers["reply-to"]).then(
+        value => value ? value.map(addr => parseDisplayName(addr)) : value
+    );
+    let subject = await getFirstHeader(full.headers["subject"]);
+    let body = getBody(full.parts);
 
-function main()
-{
+    if (!body) {
+        return null;
+    }
+
+    return {
+        header: { date, from, replyTo, subject },
+        body,
+        isPatch: (body.indexOf("\n---") >= 0 && body.indexOf("\n+++") >= 0) || body.indexOf("\ndiff --git") >= 0
+    }
+}
+
+function main() {
+    // onMessage listeners should not be async, but instead return a Promise for
+    // active responses and false if they are not responsible for the given request.
     messenger.runtime.onMessage.addListener((request, sender, sendResponse) => {
         if (request.action === "getMsg") {
-            return messenger.CopyPatch.getSelectedMessage(sender.tab.windowId);
+            return browser.messageDisplay.getDisplayedMessage(sender.tab.id).then(
+                msg => getMsgData(msg.id)
+            );
         }
         if (request.action === "clipboardWrite") {
             navigator.clipboard.writeText(request.text);
 
             messenger.messageDisplayAction.setBadgeBackgroundColor(
-                {tabId: sender.tab.id, color: "green"});
+                { tabId: sender.tab.id, color: "green" });
             messenger.messageDisplayAction.setBadgeText(
-                {tabId: sender.tab.id, text: "✔"});
+                { tabId: sender.tab.id, text: "✔" });
             setTimeout(() => {
                 messenger.messageDisplayAction.setBadgeText(
-                    {tabId: sender.tab.id, text: null});
+                    { tabId: sender.tab.id, text: null });
```

Comment:
These three are examples for style changes I was referring to.

Reply:
I will try to undo the automatic pretty format changes caused by Visual Studio Code tomorrow.